commands/build.go

```diff
@@ -643,6 +653,81 @@ func parsePrintFunc(str string) (*build.PrintFunc, error) {
 	return f, nil
 }
 
+func addGitProvenance(labels []string, contextPath string, dockerfilePath string) ([]string, error) {
+	v, ok := os.LookupEnv("BUILDX_GIT_LABELS")
+	if !ok || contextPath == "" {
+		return labels, nil
+	}
+
+	if len(labels) == 0 {
+		labels = make([]string, 0)
+	}
+
+	// figure out in which directory the git command needs to run
+	var wd string
+	if filepath.IsAbs(contextPath) {
+		wd = contextPath
+	} else {
+		cwd, _ := os.Getwd()
+		wd, _ = filepath.Abs(filepath.Join(cwd, contextPath))
+	}
+
+	// obtain Git sha of current HEAD
+	cmd := exec.Command("git", "rev-parse", "HEAD")
+	cmd.Dir = wd
+	sha, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining git head: %w", err)
+	}
+
+	// check if the current HEAD is clean
+	cmd = exec.Command("git", "diff", "--quiet", "--exit-code")
+	cmd.Dir = wd
+	_, err = cmd.Output()
+	if err == nil {
+		labels = append(labels, []string{fmt.Sprintf("%s=%s", ImageRevisionLabel, strings.TrimSpace(string(sha)))}...)
+	} else {
+		labels = append(labels, []string{fmt.Sprintf("%s=%s-dirty", ImageRevisionLabel, strings.TrimSpace(string(sha)))}...)
+	}
+
+	// add the origin url if full Git details are requested
+	if v == "full" {
+		cmd = exec.Command("git", "config", "--get", "remote.origin.url")
+		cmd.Dir = wd
+		remote, err := cmd.Output()
+		if err != nil {
+			return nil, fmt.Errorf("failed to obtain remote.origin.url: %w", err)
+		}
+		labels = append(labels, []string{fmt.Sprintf("%s=%s", ImageSourceLabel, strings.TrimSpace(string(remote)))}...)
+	}
```

Comment:
[The OCI image spec](https://github.com/opencontainers/image-spec/blob/main/annotations.md) defines the `org.opencontainers.image.source` annotation as
> URL to get source code for building the image (string)

Presumably the intention is to specify an [RFC 1738](https://www.rfc-editor.org/rfc/rfc1738) URL. However, the `remote.<name>.url` git config value is expected to be a [Git URL](https://git-scm.com/docs/git-fetch#URLS), which include the decidedly non-URL "scp-like" and local path syntaxes. Could this be a problem for consumers of this label?

What about if the build context is a local repository? Naming the clone-source remote `origin` is merely a convention. There is no guarantee that a remote of that name exists, or that the `HEAD` commit ever existed in that remote. A repository may not even have _any_ remotes configured! Even in the happy case where `HEAD` is the head of a remote-tracking branch, the remote it references may not be named `origin`.

Rather than trying to be clever and magical with introspecting the repository, and potentially getting the wrong answer, have you considered setting this label iff the build context is a remote URL, setting it to that URL?

Reply:
> [The OCI image spec](https://github.com/opencontainers/image-spec/blob/main/annotations.md) defines the `org.opencontainers.image.source` annotation as
> 
> > URL to get source code for building the image (string)
> 
> Presumably the intention is to specify an [RFC 1738](https://www.rfc-editor.org/rfc/rfc1738) URL. However, the `remote.<name>.url` git config value is expected to be a [Git URL](https://git-scm.com/docs/git-fetch#URLS), which include the decidedly non-URL "scp-like" and local path syntaxes. Could this be a problem for consumers of this label?

So far it hasn't been a problem and I kinda wanted to avoid trying to _translate_ those fetch urls into more human friendly ones. My assumption is that the value of `org.opencontainers.image.source` should be usable to get the source code for building which by using the Git urls seems correct, as one can take them and run `git clone <>`.

> What about if the build context is a local repository? Naming the clone-source remote `origin` is merely a convention. There is no guarantee that a remote of that name exists, or that the `HEAD` commit ever existed in that remote. A repository may not even have _any_ remotes configured! Even in the happy case where `HEAD` is the head of a remote-tracking branch, the remote it references may not be named `origin`.

Yes, that's a good point. I was too optimistic here. Changed that to search the remotes for one named `origin`; if not found we now take one of the configured ones. Also it won't fail now if there aren't any remotes configured for a local repository.